bokeh/cli/core.py

```diff
@@ -0,0 +1,285 @@
+        buffer (str, optional): if is `t` reads data source as string from
+            input buffer using StringIO(sys.stdin.read()) instead of
+            reading from a file or an url.
+            Default: "f"
+        sync_with_source (bool, optional): if True keep the charts source
+            created on bokeh-server sync'ed with the source acting like
+            `tail -f`.
+            Default: False
+
+
+        Attributes:
+            source (obj): datasource object for the created chart.
+            chart (obj): created chart object.
+        """
+        self.input = input_source
+        self.series = series
+        self.index = index
+        self.last_byte = -1
+        self.sync_with_source = sync_with_source
+
+        self.source = self.get_input(input_source, buffer)
+        # get the charts specified by the user
+        self.factories = create_chart_factories(chart_type)
+
+        if palette:
+            print ("Sorry, custom palettes not supported yet, coming soon!")
+
+        # define charts init parameters specified from cmd line and create chart
+        self.chart_args = get_chart_params(title, output)
+
+    def run(self):
+        """ Start the CLI logic creating the input source, data conversions,
+        chart instances to show and all other niceties provided by CLI
+        """
+        try:
+            self.chart = create_chart(
+                self.series, self.source, self.index, self.factories, **self.chart_args
+            )
+            self.chart.show()
+
+            self.has_ranged_x_axis = 'ranged_x_axis' in self.source.columns
+            self.columns = [c for c in self.source.columns if c != 'ranged_x_axis']
+
+        except TypeError:
+            if not self.series:
+                series_list = ', '.join(self.chart.values.keys())
+                print(hm.ERR_MSG_TEMPL % series_list)
+                raise
+
+        if self.sync_with_source:
+            print("animating... press ctrl-C to stop")
+            keep_source_input_sync(self.input, self.update_source, self.last_byte)
+
+    def update_source(self, new_source):
+        """ Update self.chart source with the new data retrieved from
+         new_source. It is done by parsing the new source line,
+         trasforming it to data to be appended to self.chart source
+         updating it on chart.session and actually updating chart.session
+         objects.
+
+        Args:
+            new_source (str): string that contains the new source row to
+                read to the current chart source.
+        """
+        ns = pd.read_csv(StringIO(new_source), names=self.columns)
+        len_source = len(self.source)
+
+        if self.has_ranged_x_axis:
+            ns['ranged_x_axis'] = [len_source]
+            self.index = 'ranged_x_axis'
+
+        ns.index = [len_source]
+        self.source = pd.concat([self.source, ns])
+
+        _c = create_chart(self.series, ns, self.index, self.factories, **self.chart_args)
```

Comment:
BTW I do recognize that "Wouldn't it be simple to...." are the five most dangerous words in software. :)